test/common/test-functions.js

```diff
@@ -4,12 +4,38 @@
  * These are routines used by our testing code.
  */
 
+/* Detect if we have any shadow DOM */
+window.__haveShadowDom = function() {
+    if (window.__haveShadowDomResult === undefined)
+        window.__haveShadowDomResult = Array.from(document.querySelectorAll('*')).filter(el => el.shadowRoot).length > 0;
```

Comment:
This seems pretty expensive? Likely it is more performance to write:
```
for (int x = 0 ; i < querySelector('*').length; i++ )
    if (el.shadowRoot)
         return true;
```

Alternatively:

```
let has_shadow_dom = false
document.querySelectorAll('*').forEach(element => {if (element.shadowRoot) { has_shadow_dom = true ;break; }})
```

Alternatively `document.body.contains()` would I suppose "work" but requires you to run it every time.

Reply:
Right, I rewrote it using find(). Note that this isn't super-critical, it just runs once for every frame.

I don't see how .contains() could help? that just checks if a node is a child of another node, but I don't have any child node here.